dev_tools/conftest.py

```diff
@@ -29,3 +40,85 @@ def pytest_collection_modifyitems(config, items):
     for item in items:
         if 'slow' in item.keywords:
             item.add_marker(skip_slow_marker)
+
+
+@pytest.fixture(scope="session")
+def cloned_env(testrun_uid, worker_id):
+    """Fixture to allow tests to run in a clean virtual env.
+
+    It de-duplicates installation of base packages. Assuming `virtualenv-clone` exists on the PATH,
+    it creates first a prototype environment and then clones for each new request the same env.
+    This fixture is safe to use with parallel execution, i.e. pytest-xdist. The workers synchronize
+    via a file lock, the first worker will (re)create the prototype environment, the others will
+    reuse it via cloning.
+
+    A group of tests that share the same base environment is identified by a name, `env_dir`,
+    which will become the directory within the temporary directory to hold the virtualenv.
+
+    Usage:
+
+    >>> def test_something_in_clean_env(cloned_env):
+            # base_env will point to a pathlib.Path containing the virtual env which will
+            # have quimb, jinja and whatever reqs.txt contained.
+            base_env = cloned_env("some_tests", "quimb", "jinja", "-r", "reqs.txt")
+
+            # To install new packages (that are potentially different for each test instance)
+            # just run pip install from the virtual env
+            subprocess.run(f"{base_env}/bin/pip install something".split(" "))
+            ...
+
+    Returns:
+        a function to create the cloned base environment with signature
+        `def base_env_creator(env_dir: str, *pip_install_args: str) -> Path`.
+        Use `env_dir` to specify the directory name per shared base packages.
+        Use `pip_install_args` varargs to pass arguments to `pip install`, these
+        can be requirements files, e.g. `'-r','dev_tools/.../something.txt'` or
+        actual packages as well, e.g.`'quimb'`.
+    """
+    base_dir = None
+
+    def base_env_creator(env_dir_name: str, *pip_install_args: str) -> Path:
+        """The function to create a cloned base environment."""
+        # get/create a temp directory shared by all workers
+        base_temp_path = Path(tempfile.gettempdir()) / "cirq-pytest"
+        os.makedirs(name=base_temp_path, exist_ok=True)
+        nonlocal base_dir
+        base_dir = base_temp_path / env_dir_name
+        with FileLock(str(base_dir) + ".lock"):
+            if _check_for_reuse_or_recreate(base_dir):
+                print(f"Pytest worker [{worker_id}] is reusing {base_dir} for '{env_dir_name}'.")
+            else:
+                print(f"Pytest worker [{worker_id}] is creating {base_dir} for '{env_dir_name}'.")
+                _create_base_env(base_dir, pip_install_args)
+
+        clone_dir = base_temp_path / str(uuid.uuid4())
+        shell_tools.run_cmd(*(f"virtualenv-clone {base_dir} {clone_dir}".split()))
```

Comment:
I suspect this will break if base_dir or clone_dir has spaces in its path. Fun fact: in windows 95, microsoft made the install dir `C:\Program Files` to make sure developers had to account for spaces in their paths :) 

The solution would be to 
`run_cmd('vclone', str(base_dir), str(clone_dir))` or maybe even `run_cmd('vclone', base_dir, clone_dir)` if `run_cmd` auto-stringifies its arguments

Reply:
good point :) Will change!